engine/crates/partial-caching/src/output/shapes/building.rs

```diff
@@ -0,0 +1,355 @@
+                    // TODO: This needs to be smarter.  If there's no type_condition it doesn't matter what typename
+                    // is. We also need to handle implements properly, which will require the registry.
+                    //
+                    // Will revisit later though...
+                    continue;
+                }
+
+                let defer = fragment.defer_directive();
+                if let Some(defer) = &defer {
+                    defer_stack.push(defer.label);
+                }
+
+                collect_fields(
+                    grouped_fields,
+                    defer_stack,
+                    &subset
+                        .selection_iter(fragment.selection_set())
+                        .map(|nested_selection| MergedSelection {
+                            selection: nested_selection,
+                            propagate_defer_label: selection.propagate_defer_label,
+                        })
+                        .collect::<Vec<_>>(),
+                    subset,
+                    typename,
+                );
+
+                if defer.is_some() {
+                    defer_stack.pop();
+                }
+            }
+            Selection::FragmentSpread(spread) => {
+                let Some(fragment) = spread.fragment() else { continue };
+
+                if typename != Some(fragment.type_condition()) {
+                    // TODO: This needs to be smarter.  If there's no type_condition it doesn't matter what typename
+                    // is. We also need to handle implements properly, which will require the registry.
+                    //
+                    // Will revisit later though...
+                    continue;
+                }
+
+                let defer = spread.defer_directive();
+                if let Some(defer) = &defer {
+                    defer_stack.push(defer.label);
+                }
+
+                collect_fields(
+                    grouped_fields,
+                    defer_stack,
+                    &subset
+                        .selection_iter(fragment.selection_set())
+                        .map(|nested_selection| MergedSelection {
+                            selection: nested_selection,
+                            propagate_defer_label: selection.propagate_defer_label,
+                        })
+                        .collect::<Vec<_>>(),
+                    subset,
+                    typename,
+                );
+
+                if defer.is_some() {
+                    defer_stack.pop();
+                }
+            }
+        }
+    }
+}
+
+struct MergedField<'a> {
+    response_key: &'a str,
+
+    /// The label of the defer for this selection.
+    ///
+    /// This should only be set if none of the parent fields have the same defer_label
+    defer_label: Option<&'a str>,
+
+    selections: Vec<MergedSelection<'a>>,
+}
+
+pub(super) struct MergedSelection<'a> {
```

Comment:
Yeah, `MergeSelection` emerged as a name because it's inside `MergedField` which is the output of `merge_selection_sets`.

But since the output of `merge_selection_sets` gets fed back into `collect_fields`, that needs to take a `MergedSelection` which means I need to create one even when it's not merged.

I'll have a think on a better name - not sure `MaybeDeferredSelection` is quite right either...